### DISCUSSION TITLE
`status` should not be `success` when passing a falsy query key

### DISCUSSION THREAD

**Discussion by kamranayub** [TanStack/query#426] [General] [answered]
Related to #360 and #144 

Currently when passing a falsy query key (or if the query is manual), `status` is being set to `success`. I would argue this is misleading because I _think_ the intent of a `success` status should be if the query was fetched and had data (or if it has initialData). I left a comment with some thoughts on this behavior: https://github.com/tannerlinsley/react-query/issues/144#issuecomment-621445327

The use case is just knowing whether the query was _successfully fetched_.

## Proposal

Would you be open to proposing a new status, something like `initial`, `idle`, or `ready`? That would make these kinds of checks easier without the need to check multiple things.

**States**:

- `ready` - **NEW** If the `queryFn` has not been invoked yet, such as when passed a falsy query state or in manual mode _without_ initial data.
- `success` - If the query has _received a response with no errors_ and is ready to display its data. The corresponding data property on the query is the data received from the successful fetch. If the query is `manual` _and_ has `initialData` set.
- `loading` - same as-is
- `error` - same as-is

## Current Workarounds

I think in order to _actually_ know whether a query has been invoked and was successful for sure with the way things currently are, we'd need to track `isFetching` and ensure it was `true` once before checking `status`. We could also check whether `data` is `undefined`, presumably if there was successful data, we'd have a response available (in an error, `status === 'error'` so that's fine). I don't know if that's always 100% the case since maybe a queryFn could be invoked and return `undefined` as a valid result.

**Comment by tannerlinsley** [comment, marked as answer]
- The next major version of React Query will have this missing `idle` state. It's a breaking change, so I'm not taking it lightly or rushing it.
- You can work around this by wrapping useQuery, detecting this state, and decorating the return value to have it.

**Comment by kamranayub** [comment]
Thanks, I already have a fix for our use cases, good to know! Feel free to
close or leave open until resolved in v2.

Aside: is there details anywhere about what to expect in v2? Exciting!

**Reply by tannerlinsley** [reply]
Probably not for a while :) There's not much necessitating it right now, but we'll see.
